src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,61 +101,109 @@ public T runWithAnyNode() {
     }
   }
 
-  private T runWithRetries(final int slot, int attempts, boolean tryRandomNode, JedisRedirectionException redirect) {
-    if (attempts <= 0) {
-      throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
-    }
-
-    Jedis connection = null;
-    try {
-
-      if (redirect != null) {
-        connection = this.connectionHandler.getConnectionFromNode(redirect.getTargetNode());
-        if (redirect instanceof JedisAskDataException) {
-          // TODO: Pipeline asking with the original command to make it faster....
-          connection.asking();
-        }
-      } else {
-        if (tryRandomNode) {
-          connection = connectionHandler.getConnection();
+  private T runWithRetries(final int slot) {
+    Instant deadline = Instant.now().plus(maxTotalRetriesDuration);
+
+    JedisRedirectionException redirect = null;
+    int consecutiveConnectionFailures = 0;
+    for (int attemptsLeft = this.maxAttempts; attemptsLeft > 0; attemptsLeft--) {
+      Jedis connection = null;
+      try {
+        if (redirect != null) {
+          connection = connectionHandler.getConnectionFromNode(redirect.getTargetNode());
+          if (redirect instanceof JedisAskDataException) {
+            // TODO: Pipeline asking with the original command to make it faster....
+            connection.asking();
+          }
         } else {
           connection = connectionHandler.getConnectionFromSlot(slot);
         }
+
+        return execute(connection);
+
+      } catch (JedisNoReachableClusterNodeException jnrcne) {
+        throw jnrcne;
+      } catch (JedisConnectionException jce) {
+        ++consecutiveConnectionFailures;
+        LOG.debug("Failed connecting to Redis: {}", connection, jce);
+        // "- 1" because we just did one, but the attemptsLeft counter hasn't been decremented yet
+        boolean reset = handleConnectionProblem(attemptsLeft - 1, consecutiveConnectionFailures, deadline);
+        if (reset) {
+          consecutiveConnectionFailures = 0;
+          redirect = null;
+        }
+      } catch (JedisRedirectionException jre) {
+        LOG.debug("Redirected by server to {}", jre.getTargetNode());
+        consecutiveConnectionFailures = 0;
+        redirect = jre;
+        // if MOVED redirection occurred,
+        if (jre instanceof JedisMovedDataException) {
+          // it rebuilds cluster's slot cache recommended by Redis cluster specification
+          this.connectionHandler.renewSlotCache(connection);
+        }
+      } finally {
+        releaseConnection(connection);
+      }
+      if (Instant.now().isAfter(deadline)) {
+        // TODO: change to JedisClusterOperationException or a new sub-class of it
```

Comment:
@jensgreen throwing _Max Attempts_ exception for exceeding deadline does not sound right. We either have to create new exception or use `JedisClusterOperationException`. But current user implementations may break in either case. So I'm a bit confused about what to do here.

Reply:
I see the predicament.

I can see only two ways forward that does not break current usage:
1) throw `JedisClusterMaxAttemptsException`; or
2) use a generous interpretation of `JedisClusterMaxAttemptsException` that is "I tried the maximum number of attempts allowed by the specified constraints", and throw a new exception `JedisClusterDeadlineExceededException extends JedisClusterMaxAttemptsException` that means "I tried the maximum number of attempts possible before the deadline constraint was violated".

Now that said, I think a `JedisClusterDeadlineExceededException extends JedisClusterOperationException` would be cleaner.

Which one is the way to go I think depends mostly on how you as maintainers value backwards compatibility.